### PR TITLE
feat: v2 file processing db schema

### PR DESCRIPTION
**Pull request by eeee0717**
@EurFelux 之前讨论的不同的feature有不同的endpoint，我看了paddle的文档，现在有异步解析，多个模型都已经统一url: `https://paddleocr.aistudio-app.com/api/v2/ocr/jobs`, 只需要修改model即可调用不同的解析模型,目前没看到有其他的服务商是不同endpoint有不同的url

### CODE REVIEW

File: src/main/data/migration/v2/migrators/mappings/FileProcessingOverrideMappings.ts
Comment (eeee0717) [inline src/main/data/migration/v2/migrators/mappings/FileProcessingOverrideMappings.ts:1]: @0xfullex 这里的需要创建另外的文件吗，还是应该放在`PreferenceTransformers.ts`中

**Comment by EurFelux** [comment]
> [!NOTE]
> This comment was translated by Claude.

Should we consider compatibility with paddle's old version?

---
<details>
<summary>Original Content</summary>

是否需要考虑兼容paddle旧版本？

</details>

**Comment by eeee0717** [comment]
> Should we consider compatibility with paddle's old version?

> [!NOTE]
> This comment was translated by Claude.

The keys are consistent between the new and old versions. When migrating, you only need to migrate the keys; there's no need to migrate the URL.

---
<details>
<summary>Original Content</summary>

新旧版本key是一致的，迁移的时候只需要迁移key，不用迁移url即可

</details>

**Comment by EurFelux** [comment]
> [!NOTE]
> This comment was translated by Claude.

I mean the old version of the paddleOCR service probably still uses different URLs for different services, right? Of course, we can also only support the new version, which would be more convenient for design.

---
<details>
<summary>Original Content</summary>

我的意思是paddleOCR服务的旧版本，应该还是给不同的服务配置了不同的URL吧。当然我们也可以只支持新版本，这样设计起来更方便一些。

</details>

File: packages/shared/data/presets/file-processing.ts
Comment (EurFelux) [inline packages/shared/data/presets/file-processing.ts:1]: 考虑引入zod schema验证。如果未来有可能从外部导入配置信息的话，应该会有相关字段需要经过验证。

File: packages/shared/data/presets/file-processing.ts [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+/**
+ * Feature type
+ */
+export type FileProcessorFeature = 'text_extraction' | 'markdown_conversion'
+
+/**
+ * Input type (category)
+ * Can be migrated to FileTypes enum in @types later
+ */
+export type FileProcessorInput = 'image' | 'document'
```
Comment (EurFelux) [inline packages/shared/data/presets/file-processing.ts]: 该类型被导出，且未与各个Capability进行同步。


```suggestion
export type FileProcessorInput = FeatureCapability['input']
```

File: packages/shared/data/presets/file-processing.ts [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+
+/**
+ * Output type
+ */
+export type FileProcessorOutput = 'text' | 'markdown'
```
Comment (EurFelux) [inline packages/shared/data/presets/file-processing.ts]: 同理

```suggestion
export type FileProcessorOutput = FeatureCapability['output']
```

File: packages/shared/data/presets/file-processing.ts [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+/**
+ * Feature capability definition
+ *
+ * Each capability binds a feature with its input/output and optional API settings.
+ */
+export type CapabilityMetadata = Record<string, unknown>
+
+export type TextExtractionCapability = {
+  feature: 'text_extraction'
+  input: 'image' | 'document'
```
Comment (EurFelux) [inline packages/shared/data/presets/file-processing.ts]: 当前设计下，一个Capability的input必须是单个类型，这意味着如果text extraction同时支持多种input类型，我们就必须创建多个Capability对象。这是预期的设计吗？

File: packages/shared/data/presets/file-processing.ts (hunk 1 above)
Comment (EurFelux) [inline packages/shared/data/presets/file-processing.ts]: `document`作为一种输入类型并不明确其关联了哪些类型的文档，需要jsdoc提供进一步说明。

File: packages/shared/data/presets/file-processing.ts [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+export type FileProcessorType = 'api' | 'builtin'
+
+/**
+ * Feature type
+ */
+export type FileProcessorFeature = 'text_extraction' | 'markdown_conversion'
+
+/**
+ * Input type (category)
+ * Can be migrated to FileTypes enum in @types later
```
Comment (EurFelux) [inline packages/shared/data/presets/file-processing.ts]: 该enum已迁移为`FileType`类型和`FILE_TYPE`常量对象。

**Comment by eeee0717** [comment]
> I mean the old version of the paddleOCR service probably still uses different URLs for different services, right? Of course, we can also only support the new version, which would be more convenient for design.

> [!NOTE]
> This comment was translated by Claude.

This doesn't affect anything, the key is the same. Previously, URLs were used to distinguish which model was being used, but now models are distinguished by modelId.

---
<details>
<summary>Original Content</summary>

这个不影响，key是同一个，原先是用url来区分使用的是什么模型，现在模型用modelId来区分

</details>

File: packages/shared/data/presets/file-processing.ts (hunk 1 above)
Comment (eeee0717) [reply packages/shared/data/presets/file-processing.ts]: 我觉得这里的类型是否应该等文件系统来做app的统一定义再引用会好一些？

Comment (EurFelux) [reply packages/shared/data/presets/file-processing.ts]: 可以

**Comment by eeee0717** [comment]
> [!NOTE]
> This comment was translated by Claude.

@EurFelux Are there any other places in this design that need to be changed?

---
<details>
<summary>Original Content</summary>

@EurFelux 这个设计还有需要更改的地方吗

</details>

**Comment by DeJeune** [comment]
> [!NOTE]
> This comment was translated by Claude.

It needs to be cleaned up a bit, it seems there are many merged commits.

---
<details>
<summary>Original Content</summary>

需要清理一下，看来有很多合并过的提交
</details>

**Comment by EurFelux** [comment]
> [!NOTE]
> This comment was translated by Claude.

The issue with this commit 8934fbfa0497a9587b2ea04c4244956be93bd65d is that it introduced many unrelated changes.

---
<details>
<summary>Original Content</summary>

8934fbfa0497a9587b2ea04c4244956be93bd65d 这个commit的问题，引入了很多无关改动

</details>

File: packages/shared/data/preference/preferenceTypes.ts
Comment (EurFelux) [inline packages/shared/data/preference/preferenceTypes.ts:1]: @0xfullex 我们能在这个文件里定义zod schema吗？我倾向于把schema, type guard, type放到一起

File: src/main/data/services/FileProcessingService.ts [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+  public async updateProcessor(id: FileProcessorId, updates: FileProcessorOverride): Promise<FileProcessorMerged> {
+    const overrides = this.getOverrides()
+    const nextOverrides: FileProcessorOverrides = {
+      ...overrides,
+      [id]: mergeProcessorOverrides(overrides[id], updates)
+    }
+
+    this.getPresetById(id)
+
+    await preferenceService.set('file_processing.overrides', nextOverrides)
```
Comment (DeJeune) [inline src/main/data/services/FileProcessingService.ts]: **Bug:** `getPresetById(id)` is called *after* `preferenceService.set()`. If the processor ID doesn't exist in presets, the preference is already written with invalid data before the `notFound` error is thrown.

Move the validation before the write:

```ts
public async updateProcessor(id: FileProcessorId, updates: FileProcessorOverride): Promise<FileProcessorMerged> {
  this.getPresetById(id) // validate first
  const overrides = this.getOverrides()
  const nextOverrides: FileProcessorOverrides = {
    ...overrides,
    [id]: mergeProcessorOverrides(overrides[id], updates)
  }

  await preferenceService.set('file_processing.overrides', nextOverrides)
  // ...
}
```

File: src/main/data/services/FileProcessingService.ts [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+): FileProcessorOverride {
+  return {
+    ...current,
+    ...updates,
+    capabilities: mergeCapabilityOverrides(current?.capabilities, updates?.capabilities),
+    options: {
+      ...current?.options,
+      ...updates?.options
+    }
+  }
```
Comment (DeJeune) [inline src/main/data/services/FileProcessingService.ts:81]: **Nit:** `mergeProcessorOverrides` always produces `options: {}` even when both `current` and `updates` have no options, since spreading two undefineds into an object yields `{}`. This means every update will persist an empty `options` field unnecessarily.

Consider guarding:

```ts
function mergeProcessorOverrides(
  current?: FileProcessorOverride,
  updates?: FileProcessorOverride
): FileProcessorOverride {
  const mergedOptions = current?.options || updates?.options
    ? { ...current?.options, ...updates?.options }
    : undefined

  return {
    ...current,
    ...updates,
    capabilities: mergeCapabilityOverrides(current?.capabilities, updates?.capabilities),
    ...(mergedOptions !== undefined && { options: mergedOptions })
  }
}
```

File: packages/shared/data/presets/file-processing.ts [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+export type FeatureCapability = z.infer<typeof FeatureCapabilitySchema>
+
+/**
+ * Input type (category)
+ * Derived from FeatureCapability to keep definitions in sync.
+ */
+export type FileProcessorInput = FeatureCapability['inputs'][number]
+
+/**
+ * Output type
```
Comment (DeJeune) [inline packages/shared/data/presets/file-processing.ts:102]: **Nit:** `FileProcessorMetadataSchema` uses `z.record(z.string(), z.never())` which only allows `{}`, while `CapabilityMetadataSchema` on line 113 uses `z.record(z.string(), z.unknown())` which allows anything. This inconsistency is confusing — the comment says "reserved for future use" but one schema is locked down and the other is wide open. Consider aligning them or adding a brief comment explaining the intentional difference.

File: packages/shared/data/presets/file-processing.ts [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+        feature: 'markdown_conversion',
+        inputs: ['document'],
+        output: 'markdown',
+        apiHost: 'http://127.0.0.1:8000'
+      }
+    ]
+  }
+] as const satisfies readonly FileProcessorTemplate[]
+
+void FileProcessorTemplatesSchema.parse(PRESETS_FILE_PROCESSORS)
```
Comment (DeJeune) [inline packages/shared/data/presets/file-processing.ts]: **Significant:** This `void FileProcessorTemplatesSchema.parse(...)` runs at module import time and will crash the entire app at startup if validation fails. While the intent (catching preset data bugs early) is good, this is risky in production. Consider:

1. Wrapping in a try/catch that logs the error but doesn't crash, or
2. Moving this to a test-only assertion (there's already a test for this: `'validates built-in presets'`), or
3. At minimum, adding a comment explaining the intentional crash-on-import behavior so future contributors don't accidentally break startup.

File: packages/shared/data/presets/file-processing.ts [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+            useDocUnwarping: false
+          }
+        }
+      },
+      {
+        feature: 'markdown_conversion',
+        inputs: ['document'],
+        output: 'markdown',
+        apiHost: 'https://paddleocr.aistudio-app.com/',
+        modelId: 'PaddleOCR-VL-1.5',
```
Comment (DeJeune) [inline packages/shared/data/presets/file-processing.ts]: **Significant:** The `FileProcessorMergedSchema` extends `FileProcessorTemplateSchema` which inherits the `.strict()` constraint. However, the merged type adds `apiKeys` and `options` via `.extend()` — that's fine. But the capabilities array still uses the strict `FeatureCapabilitySchema` from the template.

When `mergeProcessorConfig` in `FileProcessingService` spreads override fields into capabilities:
```ts
capabilities: preset.capabilities.map((capability) => ({
  ...capability,
  ...override?.capabilities?.[capability.feature]
}))
```

The spread can add extra fields (from `CapabilityOverride`) that aren't defined in the strict capability schema. This means the merged result would fail if re-validated against `FileProcessorMergedSchema`. Since you're not re-validating the merged output this works at runtime, but it's a type-level inconsistency worth noting. Consider either:
- Relaxing the capability schema in the merged variant, or
- Explicitly picking only known fields from the override when merging

File: packages/shared/data/preference/preferenceTypes.ts [hunk 10]
```diff
@@ -105,6 +105,44 @@ export type MultiModelMessageStyle = 'horizontal' | 'vertical' | 'fold' | 'grid'
...
+
+export type FileProcessorCapabilityOverrides = Partial<Record<FileProcessorFeature, CapabilityOverride>>
+
+export type FileProcessorOverride = {
+  apiKeys?: string[]
+  capabilities?: FileProcessorCapabilityOverrides
+  options?: FileProcessorOptions
+}
+
+export type FileProcessorOverrides = Partial<Record<FileProcessorId, FileProcessorOverride>>
```
Comment (DeJeune) [inline packages/shared/data/preference/preferenceTypes.ts]: **Nit:** Agreeing with @EurFelux's earlier comment — since there are already Zod schemas defined in `file-processing.ts` for these same types (e.g., `FileProcessorOverrideSchema`, `CapabilityOverrideSchema`), having plain TypeScript type definitions here creates a dual-source-of-truth risk. If either side drifts, the types will silently diverge. Consider deriving these types from the Zod schemas using \`z.infer\` (or vice versa) to keep them in sync, similar to how \`FileProcessorTemplate\` is already done.

**Review by DeJeune — COMMENTED** [review]
## Review Summary

Well-structured PR that establishes the file processing data layer for v2. The schema design (templates + overrides), migration logic, and service layer are well thought out. Good test coverage across all new modules.

### Critical / Bug (1)
- **`updateProcessor` validation order** — `getPresetById()` is called after `preferenceService.set()`, meaning invalid processor IDs corrupt preferences before the error is thrown.

### Significant (2)
- **Runtime parse at import time** — `void FileProcessorTemplatesSchema.parse(PRESETS_FILE_PROCESSORS)` will crash the app at startup if validation fails. There's already a test for this; consider making the import-time assertion softer or at least documenting the intent.
- **Merged schema strict mode mismatch** — `FileProcessorMergedSchema` inherits strict capability schemas from the template, but `mergeProcessorConfig` spreads override fields that aren't in the strict schema. This would fail re-validation.

### Minor / Nit (3)
- **Empty `options: {}` on every merge** — `mergeProcessorOverrides` always creates an empty options object even when neither side has options.
- **Metadata schema inconsistency** — `FileProcessorMetadataSchema` (`z.never()` values) vs `CapabilityMetadataSchema` (`z.unknown()` values) is confusing without explanation.
- **Dual source of truth** — Plain TS types in `preferenceTypes.ts` duplicate Zod schemas in `file-processing.ts`; consider deriving one from the other.

### Positives
- Clean separation of concerns: templates (read-only presets) vs overrides (user config) vs merged (API responses)
- Thorough migration logic with good edge case handling (paddleocr special-casing, preset default skipping, empty pruning)
- Solid test coverage for migration, service, and schema validation
- Good use of discriminated unions for capability types
- CI is green ✅

File: packages/shared/data/preference/preferenceTypes.ts (hunk 10 above)
Comment (eeee0717) [reply packages/shared/data/preference/preferenceTypes.ts]: 已处理。

按照 v2 这边的约定，我保留了 `preferenceTypes.ts` 作为 source of truth，没有让 `preferenceTypes` 反向依赖 `file-processing.ts`。

这次不是继续从 preset schema 反推 preference 类型，而是改成 `file-processing.ts` 从 `preferenceTypes.ts` 引入这些 preference-facing 类型，再通过 `z.ZodType<...>` 把 Zod schema 和这些类型绑定起来。这样既能保证 schema 和 preference 类型保持一致，也不会引入 `preferenceTypes -> presets` 这种反向依赖。

另外我也顺手收紧了 `mergeProcessorConfig` 的逻辑，merged capability 现在只会合并 schema 允许的字段（`apiHost`、`modelId`、`metadata`），不会再把未知 override 字段直接透传到返回结果里。对应的回归测试也补上了，确保未知字段不会泄漏到 merged config。

File: packages/shared/data/presets/file-processing.ts [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+ */
+export const FileProcessorInputSchema = FileTypeSchema.extract([FILE_TYPE.IMAGE, FILE_TYPE.DOCUMENT])
+
+const FileProcessorTextOutputSchema = FileTypeSchema.extract([FILE_TYPE.TEXT])
+
+/**
+ * Output content format schema
+ * `text` reuses the canonical file type, while `markdown` remains a processing-specific format.
+ */
+export const FileProcessorOutputSchema = z.union([FileProcessorTextOutputSchema, z.literal('markdown')])
```
Comment (DeJeune) [inline packages/shared/data/presets/file-processing.ts:55]: **Nit:** `FileProcessorMetadataSchema` is defined here with `z.record(z.string(), z.never())` (which rejects all keys), but it is only used in `FileProcessorTemplateSchema` as `.optional()`. Since `.strict()` on the template already forbids extra keys, this schema effectively does nothing useful — any value that passes `z.never()` is vacuously empty `{}`.

Consider removing `FileProcessorMetadataSchema` entirely and the `metadata` field from the template schema, or replacing `z.never()` with a more permissive value type if you intend to allow metadata in the future. The current definition is confusing because it *looks* like it supports metadata but actually forbids all content.

File: src/main/data/services/FileProcessingService.ts [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+    ...(override?.metadata !== undefined ? { metadata: override.metadata } : {})
+  }
+}
+
+export class FileProcessingService {
+  private static instance: FileProcessingService
+
+  private constructor() {}
+
+  public static getInstance(): FileProcessingService {
```
Comment (DeJeune) [inline src/main/data/services/FileProcessingService.ts:89]: **Bug:** The `getPresetById` call on line 80 validates the processor exists but happens *after* the overrides have already been computed and before they are persisted. This means if an invalid ID somehow passes the TypeScript type guard, the overrides object would already be constructed with the invalid entry.

More importantly, since `id` is typed as `FileProcessorId`, the runtime check in `getPresetById` will always pass for valid preset IDs. However, the order is still semantically wrong — you should validate the ID exists *before* computing the merged overrides:

```typescript
public async updateProcessor(id: FileProcessorId, updates: FileProcessorOverride): Promise<FileProcessorMerged> {
  this.getPresetById(id) // validate first
  const overrides = this.getOverrides()
  const nextOverrides: FileProcessorOverrides = {
    ...overrides,
    [id]: mergeProcessorOverrides(overrides[id], updates)
  }
  await preferenceService.set('file_processing.overrides', nextOverrides)
  // ...
}
```

File: src/main/data/services/FileProcessingService.ts [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+      hasApiKeys: Boolean(nextOverrides[id]?.apiKeys?.length),
+      capabilityCount: Object.keys(nextOverrides[id]?.capabilities || {}).length
+    })
+
+    return this.mergeProcessorConfig(id, nextOverrides)
+  }
+
+  private getOverrides(): FileProcessorOverrides {
+    return preferenceService.get('file_processing.overrides') ?? {}
+  }
```
Comment (DeJeune) [inline src/main/data/services/FileProcessingService.ts:129]: **Significant:** The capability merge via spread (`...override?.capabilities?.[capability.feature]`) is a shallow merge that silently overwrites the template's `apiHost`/`modelId`/`metadata` with the override values. This is correct for `apiHost` and `modelId`, but for `metadata` it does a full replacement instead of a deep merge.

For example, if the template has `metadata: { optionalPayload: { useDocOrientationClassify: false, useDocUnwarping: false } }` and the user overrides `metadata: { optionalPayload: { useDocOrientationClassify: true } }`, the resulting metadata will lose `useDocUnwarping`.

Is this the intended behavior? If metadata should be deep-merged, consider using a deep merge utility here. If full replacement is intended, it would be worth documenting that design choice.

File: src/main/data/api/handlers/fileProcessing.ts [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+    }
+  },
+
+  '/file-processing/processors/:id': {
+    GET: async ({ params }) => {
+      return await fileProcessingService.getProcessorById(params.id)
+    },
+
+    PATCH: async ({ params, body }) => {
+      return await fileProcessingService.updateProcessor(params.id, body)
```
Comment (DeJeune) [inline src/main/data/api/handlers/fileProcessing.ts]: **Significant:** The PATCH handler passes `body` directly to `updateProcessor` without any validation. Since this is a system boundary (IPC from renderer), the `body` should be validated against `FileProcessorOverrideSchema` before being used.

Without validation, malformed or unexpected data from the renderer could corrupt the stored preferences.

```typescript
PATCH: async ({ params, body }) => {
  const validated = FileProcessorOverrideSchema.parse(body)
  return await fileProcessingService.updateProcessor(params.id, validated)
}
```

File: packages/shared/data/presets/file-processing.ts [hunk 15]
```diff
@@ -0,0 +1,342 @@
...
+ * Processor template (read-only metadata)
+ *
+ * Note: Display name is retrieved via i18n key `processor.${id}.name`
+ */
+export const FileProcessorTemplateSchema = z
+  .object({
+    id: FileProcessorIdSchema,
+    type: FileProcessorTypeSchema,
+    capabilities: z.array(FeatureCapabilitySchema).min(1)
+  })
```
Comment (DeJeune) [inline packages/shared/data/presets/file-processing.ts]: **Nit:** The `FileProcessorTemplateSchema` includes `metadata: FileProcessorMetadataSchema.optional()` on line 119 in the test, but looking at the actual schema definition (line 115-122), the `metadata` field was removed from the template schema while `FileProcessorMetadataSchema` is still defined and exported (line 53). This is good — but the `FileProcessorMetadataSchema` export is now dead code since no consumer uses it. Consider removing lines 53-54 to keep the module clean.

File: packages/shared/data/presets/file-processing.ts [hunk 16]
```diff
@@ -0,0 +1,342 @@
...
+    capabilities: [
+      {
+        feature: 'markdown_conversion',
+        inputs: ['document'],
+        output: 'markdown',
+        apiHost: 'http://127.0.0.1:8000'
+      }
+    ]
+  }
+] as const satisfies readonly FileProcessorTemplate[]
```
Comment (DeJeune) [inline packages/shared/data/presets/file-processing.ts]: **Nit:** This runtime parse call (`void FileProcessorTemplatesSchema.parse(...)`) runs at module load time and will throw if the presets are invalid. While useful for development, this is an unconditional side effect on import that could crash the app at startup if there is a schema mismatch.

The test already covers this (`it("validates built-in presets", ...)`) so this line is arguably redundant. Consider removing it or wrapping in a try/catch with proper error logging.

File: src/main/data/migration/v2/migrators/mappings/FileProcessingOverrideMappings.ts [hunk 17]
```diff
@@ -0,0 +1,240 @@
...
+
+function isRecord(value: unknown): value is Record<string, unknown> {
+  return typeof value === 'object' && value !== null && !Array.isArray(value)
+}
+
+function isFileProcessorId(value: unknown): value is FileProcessorId {
+  return typeof value === 'string' && FILE_PROCESSOR_IDS.includes(value as FileProcessorId)
+}
+
+function ensureOverride(overrides: FileProcessorOverrides, id: FileProcessorId): FileProcessorOverride {
```
Comment (DeJeune) [inline src/main/data/migration/v2/migrators/mappings/FileProcessingOverrideMappings.ts:28]: **Nit:** The cast `Object.keys(updates ?? {}) as FileProcessorFeature[]` assumes all keys of the `updates` object are valid `FileProcessorFeature` values. Since this is migration code processing legacy data, the keys could theoretically be anything. Consider adding a guard:

```typescript
for (const key of Object.keys(updates ?? {})) {
  if (!FILE_PROCESSOR_FEATURES.includes(key as FileProcessorFeature)) continue
  const feature = key as FileProcessorFeature
  merged[feature] = { ...current?.[feature], ...updates?.[feature] }
}
```

This would make the migration more resilient to unexpected legacy data.

**Review by DeJeune — COMMENTED** [review]
## Review Summary

Overall this is a well-structured PR that cleanly introduces the file processing data layer with good separation of concerns (templates vs overrides vs merged config). The Zod schema design, discriminated union for capabilities, and the migration logic are thoughtfully designed.

### Significant Issues

1. **Missing input validation at PATCH handler** — The `body` from renderer is passed directly to `updateProcessor` without schema validation. This is a system boundary and should validate input.

2. **Shallow metadata merge** — The capability merge in `mergeProcessorConfig` does a shallow spread, which means nested `metadata` objects (like `optionalPayload`) will be fully replaced rather than deep-merged. Please confirm if this is intentional.

### Minor Issues

3. **Validation order in `updateProcessor`** — `getPresetById(id)` is called after computing the merged overrides. Semantically it should validate first.

4. **Dead code** — `FileProcessorMetadataSchema` (lines 53-54) is exported but unused since the `metadata` field was removed from `FileProcessorTemplateSchema`.

5. **Runtime parse side effect** — `void FileProcessorTemplatesSchema.parse(PRESETS_FILE_PROCESSORS)` runs at import time; the test already covers this validation.

6. **Unsafe cast in migration code** — `Object.keys(updates) as FileProcessorFeature[]` could benefit from a runtime guard since migration data may be unexpected.

### Positives

- Clean template/override/merged architecture with clear data flow
- Good use of Zod discriminated unions for capability types
- Thorough migration logic with proper handling of PaddleOCR special case (skipping apiHost migration)
- Comprehensive test coverage across schemas, service, and migration
- All CI checks passing

File: packages/shared/data/presets/file-processing.ts [hunk 18]
```diff
@@ -0,0 +1,342 @@
...
+ * `text` reuses the canonical file type, while `markdown` remains a processing-specific format.
+ */
+export const FileProcessorOutputSchema = z.union([FileProcessorTextOutputSchema, z.literal('markdown')])
+
+/**
+ * Feature capability definition
+ *
+ * Each capability binds a feature with its supported inputs, output, and optional API settings.
+ */
+export const CapabilityMetadataSchema = z.record(z.string(), z.unknown())
```
Comment (DeJeune) [inline packages/shared/data/presets/file-processing.ts]: 这里命名明确以下是FileProcess的，免得和Model冲突

File: packages/shared/data/preference/preferenceTypes.ts [hunk 19]
```diff
@@ -105,6 +105,44 @@ export type MultiModelMessageStyle = 'horizontal' | 'vertical' | 'fold' | 'grid'
...
+  'doc2x',
+  'mistral',
+  'open-mineru'
+] as const
+
+export type FileProcessorId = (typeof FILE_PROCESSOR_IDS)[number]
+
+export type FileProcessorOptions = Record<string, unknown>
+
+export type CapabilityOverride = {
```
Comment (DeJeune) [inline packages/shared/data/preference/preferenceTypes.ts]:
```suggestion
export type FileProcessorCapabilityOverride = {
```

File: packages/shared/data/preference/preferenceSchemas.ts [hunk 20]
```diff
@@ -404,6 +404,12 @@ export interface PreferenceSchemas {
     'feature.translate.model_prompt': string
     // redux/settings/targetLanguage
     'feature.translate.target_language': string
+    // redux/preprocess/defaultProvider
+    'file_processing.default.markdown_conversion': PreferenceTypes.FileProcessorId
```
Comment (0xfullex) [inline packages/shared/data/preference/preferenceSchemas.ts]: 放feature. 下

Comment (eeee0717) [reply packages/shared/data/preference/preferenceSchemas.ts]: fix

**Review by DeJeune — APPROVED** [review]
LGTM

File: packages/shared/data/preference/preferenceSchemas.ts [hunk 21]
```diff
@@ -634,6 +640,9 @@ export const DefaultPreferences: PreferenceSchemas = {
     'feature.csaas.enabled': false,
     'feature.csaas.host': '127.0.0.1',
     'feature.csaas.port': 23333,
+    'feature.file_processing.default.markdown_conversion': 'mineru',
```
Comment (0xfullex) [inline packages/shared/data/preference/preferenceSchemas.ts]: default不用单独作为一个节点吧，尽量平铺，default_xxx_xxx

**Review by 0xfullex — CHANGES_REQUESTED** [review]
# PR #13328 Comprehensive Review

## V2 Data Architecture Compliance: PASS

The PR correctly follows the v2 Layered Preset Pattern:
- Presets in `packages/shared/data/presets/file-processing.ts` with proper naming (`PRESETS_FILE_PROCESSORS`)
- User overrides stored as Preference (`feature.file_processing.overrides`) -- correct system choice
- Handler -> Service architecture matches existing patterns (TopicService, MessageService)
- API routes follow REST conventions, logging uses `loggerService`, preference keys follow dot-notation
- Migration follows established complex mapping pattern (matches `migrateWebSearchProviders`)
- i18n display names via `processor.${id}.name` -- cleaner than WebSearch pattern

---

## Critical Issues (Must Fix Before Merge)

### C1. `metadata` Three-Way Inconsistency

The `metadata` field has contradictory treatment across three layers:

| Location | Status |
|----------|--------|
| Template capability schema (`file-processing.ts`) | **Removed** (commit `47ea921`) |
| `CapabilityOverride` type (`preferenceTypes.ts`) | **Still present** |
| `mergeCapabilityConfig` function (`FileProcessingService.ts`) | **Still merges metadata** |

If metadata was intentionally removed from templates, it should also be removed from:
1. `CapabilityOverride.metadata` in `preferenceTypes.ts`
2. `FileProcessorCapabilityOverrideSchema` in `file-processing.ts`
3. `mergeCapabilityConfig` function in `FileProcessingService.ts`

If metadata IS needed for overrides, then `FileProcessorMergedSchema` must allow it -- otherwise re-validating a merged processor against the strict schema would fail.

---

## Important Issues (Should Fix)

### I1. `CapabilityOverride` Naming Inconsistency
**File**: `packages/shared/data/preference/preferenceTypes.ts`

The type is exported as `CapabilityOverride` -- too generic. All other types in this file use the `FileProcessor` prefix (`FileProcessorType`, `FileProcessorFeature`, `FileProcessorOverride`, etc.). Previously flagged by @DeJeune.

**Fix**: Rename to `FileProcessorCapabilityOverride` and update all references.

### I2. Unsafe Type Cast in `mergeCapabilityOverrides`
**File**: `src/main/data/services/FileProcessingService.ts`

```typescript
for (const feature of Object.keys(updates ?? {}) as FileProcessorFeature[])
```

Casts arbitrary object keys to `FileProcessorFeature[]`. If stored preferences contain unexpected keys (version downgrade, corruption), those keys pass through unchecked.

**Fix**: Add runtime guard:
```typescript
for (const key of Object.keys(updates ?? {})) {
  if (!FILE_PROCESSOR_FEATURES.includes(key as FileProcessorFeature)) continue
  const feature = key as FileProcessorFeature
  // ...
}
```

### I3. Silent Skip of Unknown Providers During Migration (No Logging)
**File**: `src/main/data/migration/v2/migrators/mappings/FileProcessingOverrideMappings.ts`

Both `mergePreprocessProvider` and `mergeOcrProvider` silently `return` when `isFileProcessorId()` returns false. Users with custom/plugin providers lose their entire config (including API keys) with no log trace.

**Fix**: Add `logger.warn` when a provider is skipped due to unknown ID.

### I4. `normalizeLangs` Silent Empty Return for Unexpected Types
**File**: `FileProcessingOverrideMappings.ts`

When input is neither array nor record, returns `[]` silently -- language config is dropped without trace.

**Fix**: Add `logger.warn` for the unexpected-type fallthrough branch.

### I5. Non-English Description in `target-key-definitions.json`
**File**: `v2-refactor-temp/tools/data-classify/data/target-key-definitions.json`

Contains Chinese text: `"复杂迁移：合并 redux/preprocess.providers 与 redux/ocr.providers 为 file processing overrides"`. Should be English per project conventions.

---

## Suggestions (Nice to Have)

1. **Zod parse errors in PATCH handler surface as 500 instead of 422** -- Use `safeParse` + `DataApiErrorFactory.validation()` instead of `.parse()` in `fileProcessing.ts` handler
2. **No tests for API handlers** -- `src/main/data/api/handlers/fileProcessing.ts` has zero test coverage (PATCH validation, error propagation)
3. **Missing migration edge case tests** -- empty/missing sources, unknown processor IDs, OCR `config.api` nested path, duplicate API key dedup
4. **`FileProcessorOptions = Record<string, unknown>` is extremely loose** -- Consider introducing processor-specific option types as tech debt
5. **`apiKeys` replacement vs append behavior not tested** -- `updateProcessor` replaces apiKeys (spread), should be tested explicitly

---

## Strengths

- Discriminated union for `FileProcessorFeatureCapability` is excellently designed -- makes illegal states unrepresentable
- Const-derived ID types maintain single source of truth
- Migration logic is thorough with proper edge handling (PaddleOCR special-casing, preset default skipping, empty pruning)
- Solid test coverage: 507 new lines across 5 test files
- Forward-compatibility test for unknown capability override fields
- Clean separation of concerns across all layers

---

**Summary**: 1 critical + 5 important issues. Please resolve C1 (metadata inconsistency) and I1-I2 (naming + unsafe cast) before merge. I3-I5 are recommended improvements.

**Comment by eeee0717** [comment]
@0xfullex addressed in 4b4bf97d5.

Handled in this round:
- C1: removed file processor capability metadata from presets, override types/schema, and service merge logic; added regression coverage so legacy stored metadata is ignored.
- I1: renamed `CapabilityOverride` to `FileProcessorCapabilityOverride`.
- I2: added runtime feature-key guards in `mergeCapabilityOverrides()` so invalid keys are dropped instead of being trusted via cast.
- I3: added `logger.warn` when unknown preprocess/OCR providers are skipped during migration.
- I4: added `logger.warn` when OCR `langs` has an invalid runtime type.
- I5: translated the `target-key-definitions.json` description to English.

Also covered from the suggestion list:
- PATCH handler now uses `safeParse + DataApiErrorFactory.validation()` so invalid bodies surface as 422 instead of 500.
- Added file processing handler tests for validation and error propagation.
- Added migration edge-case tests for missing sources, nested OCR `config.api`, unknown provider IDs, invalid `langs`, and API key dedupe.
- Added an explicit test locking in current `apiKeys` replacement semantics.

The only remaining nice-to-have from the summary is tightening `FileProcessorOptions` beyond `Record<string, unknown>`, which I have not changed in this patch.

**Review by 0xfullex — APPROVED** [review]
## Final Review — All Prior Issues Resolved

Verified all review comments from @EurFelux, @DeJeune, and my previous review (CHANGES_REQUESTED) have been properly addressed:

### @0xfullex Prior Review Items
- **C1 metadata inconsistency**: Fully resolved — metadata cleanly removed from `FileProcessorCapabilityOverride` type, override schema (`.strict()`), and `mergeCapabilityConfig`
- **I1 naming**: Renamed to `FileProcessorCapabilityOverride`
- **I2 unsafe cast**: `isFileProcessorFeature()` runtime guard added in `mergeCapabilityOverrides`
- **I3/I4 silent skips**: `logger.warn` added for unknown providers and invalid langs
- **I5 non-English text**: Description now in English
- **Inline comments**: Preference keys flattened to `feature.file_processing.default_markdown_conversion` / `default_text_extraction`

### @DeJeune Prior Review Items
- **Bug (validation order)**: `getPresetById(id)` now called before any override computation
- **PATCH handler validation**: `safeParse` + `DataApiErrorFactory.validation()` with structured field errors and handler tests
- **Runtime parse side effect**: Removed from production code
- **Merged schema strict mode**: `mergeCapabilityConfig` now explicitly picks only `apiHost`/`modelId`; regression tests confirm unknown fields are dropped
- **Empty options**: Conditional merging; test confirms no empty `options: {}` persisted
- **Dead code**: `FileProcessorMetadataSchema` removed
- **Dual source of truth**: `preferenceTypes.ts` as source, schemas use `z.ZodType<...>` binding

### Quality
- 916 new test lines across 8 test files
- Clean template/override/merged architecture
- Good migration resilience with logging for edge cases
- CI green (basic-checks, general-test pass)

LGTM
